zengjiaxing/src/App.js

```diff
@@ -0,0 +1,345 @@
+import React, { Component } from 'react';
+import './App.css';
+import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
+import Paper from 'material-ui/Paper';
+import TextField from 'material-ui/TextField';
+import RaisedButton from 'material-ui/RaisedButton';
+import Checkbox from 'material-ui/Checkbox';
+import Chip from 'material-ui/Chip';
+import Popover, {PopoverAnimationVertical} from 'material-ui/Popover';
+import Menu from 'material-ui/Menu';
+import MenuItem from 'material-ui/MenuItem';
+import IconButton from 'material-ui/IconButton';
+import ActionDeleteForever from 'material-ui/svg-icons/action/delete-forever';
+import EditorModeEdit from 'material-ui/svg-icons/editor/mode-edit';
+
+import TaskEditDialog from './TaskEditDialog/TaskEditDialog';
+
+class App extends Component {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            search_mode: false,
+
+            filter_menu_open: false,
+            filter: 'All',
+
+            search_menu_open: false,
+            input_type: 'Add',
+
+            todo_list: [],
+            display_list: [],
+            task_input: ''
+        };
+    }
+
+
+    //主要功能部分========================================================
+    handleInputText = (event) => {
+        const input = event.target.value;
+        this.setState({task_input: input});
+        if(this.state.input_type !== 'Add') {
+            const { todo_list } = this.state;
+            let display_list = [];
+            switch (this.state.input_type) {
+                case 'Tag':
+                    display_list = todo_list.filter((item) => item.tag.search(input) !== -1);
+                    break;
+                case 'Content':
+                    display_list = todo_list.filter((item) => item.content.search(input) !== -1);
+                    break;
+                default: break;
+            }
+            this.setState({display_list});
+        }
+    };
+
+    handleAddTask = (content) => {
+        let { todo_list } = this.state;
+        todo_list.unshift({
```

Comment:
https://medium.com/@housecor/handling-state-in-react-four-immutable-approaches-to-consider-d1f5c00249d5